### PR TITLE
feat(document-preprocessor): add text extraction for plain text and Markdown files

### PR DESCRIPTION
**Pull request by shige**
## Summary
Add extractText function to support text extraction from plain text and Markdown files with configurable encoding options. This complements the existing PDF extraction capabilities.

## Related Issue
part of #1827 

## Changes

- Add extractText function supporting Buffer, Uint8Array, and ArrayBuffer inputs
- Support configurable encoding (defaults to utf-8) and abort signal
- Add comprehensive test coverage including edge cases and different input types
- Export TextExtractionOptions and TextExtractionResult types from index
- Update README with usage examples and complete API documentation

## Other Information
An implementation using this package will be created in the next PR.

___

### **PR Type**
Enhancement

___

### **Description**
- Add text extraction for plain text and Markdown files

- Support configurable encoding options with utf-8 default

- Include comprehensive test coverage for various input types

- Export new types and update documentation

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Binary Input"] --> B["extractText Function"]
  B --> C["Buffer Conversion"]
  C --> D["Text Decoding"]
  D --> E["Text Result"]
  F["Options"] --> B
  F --> G["Encoding"]
  F --> H["Abort Signal"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Add text extraction exports</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/document-preprocessor/src/index.ts

<ul><li>Export <code>extractText</code> function from text module<br> <li> Export <code>TextExtractionOptions</code> and <code>TextExtractionResult</code> types</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1889/files#diff-bc748d842d5a3454aaac7eab6a20fe343c605b91808962b7989b224d9d0c7018">+3/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>text.ts</strong><dd><code>Implement text extraction functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/document-preprocessor/src/text.ts

<ul><li>Implement <code>extractText</code> function for plain text and markdown<br> <li> Support Buffer, Uint8Array, and ArrayBuffer inputs<br> <li> Handle configurable encoding and abort signals</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1889/files#diff-a9610ca6d7f5a4e359bf55db621aa7764d5d9a411a552b6f953c9416c38e0e4a">+36/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>types.ts</strong><dd><code>Define text extraction type interfaces</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/document-preprocessor/src/types.ts

<ul><li>Add <code>TextExtractionOptions</code> interface with encoding and signal<br> <li> Add <code>TextExtractionResult</code> interface with text property</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1889/files#diff-a52f28b2f1751e5fee3b0ae28bdf7b1b3503fdaa46740cbc665f8dd431581b26">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>text.test.ts</strong><dd><code>Add text extraction test coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/document-preprocessor/src/text.test.ts

<ul><li>Add comprehensive test suite for <code>extractText</code> function<br> <li> Test plain text, markdown, empty content, and encoding scenarios<br> <li> Include abort signal and Uint8Array input tests</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1889/files#diff-4b241d933c6699114f955993ec7aff83a3093ea957466ccaac27b343ce42cecc">+57/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Update documentation for text extraction</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/document-preprocessor/README.md

<ul><li>Update package description to include text and markdown support<br> <li> Add usage examples for text extraction<br> <li> Document complete API with type exports<br> <li> Reorganize features section with clear categorization</ul>

</details>

  </td>
  <td><a href="https://github.com/giselles-ai/giselle/pull/1889/files#diff-5132204a291206bcc463b7b6a88a675afa40f8c6595c3f2b14d86856b331d2fa">+59/-5</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * Added text extraction for plain text and Markdown alongside existing PDF support.
  * Supports custom encodings and request cancellation.
* **Documentation**
  * Updated README with expanded feature descriptions and a new usage example for plain text/Markdown.
  * Refreshed API overview to include the new text extraction method and clarified type exports.
* **Tests**
  * Added comprehensive tests covering various inputs, encodings, empty content, cancellation, and Uint8Array handling.

---

> [!NOTE]
> Introduce `extractText` for plain text/Markdown with encoding and abort support, plus new types, tests, and documentation updates.
> 
> - **Core**
>   - Add `extractText(input, options?)` for plain text/Markdown with configurable encoding and `AbortSignal` support (`src/text.ts`).
> - **Types/Exports**
>   - Define `TextExtractionOptions`, `TextExtractionResult` and export via `src/index.ts`.
> - **Tests**
>   - Add `src/text.test.ts` covering multiple inputs, encodings, empty content, and abort behavior.
> - **Docs**
>   - Update `README.md` with text/Markdown support, usage example, expanded API table, and type exports.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 9630365f53987d90d95d288464e4e124655ed803. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by shige** [comment]
@coderabbitai review

**Comment by shige** [comment]
@codex review

File: packages/document-preprocessor/src/text.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	let buffer: Buffer;
+	if (Buffer.isBuffer(input)) {
+		buffer = input;
+	} else if (input instanceof Uint8Array) {
+		buffer = Buffer.from(input);
+	} else if (ArrayBuffer.isView(input)) {
+		buffer = Buffer.from(input.buffer, input.byteOffset, input.byteLength);
+	} else {
+		buffer = Buffer.from(input);
+	}
```
[the remark this answers is not in the document]
Comment (shige) [reply packages/document-preprocessor/src/text.ts:31]: fixed cddbcc4582e9fa3fe070a265d89dd6f7e5744c25

**Review by toyamarinyon — APPROVED** [review]

**Comment by shige** [comment]
The base commit has been adjusted due to adjustments made to the migration file at https://github.com/giselles-ai/giselle/pull/1885.
The commit for this PR has been cherry-picked, so there are no changes to the modifications.

**Comment by shige** [comment]
Thank you for reviewing! 🚀
